lc0/src/neural/network_opencl.cc

```diff
@@ -0,0 +1,447 @@
+ 
+ You should have received a copy of the GNU General Public License
+ along with Leela Chess.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "neural/factory.h"
+#include "neural/network.h"
+
+#include <condition_variable>
+#include <cassert>
+#include <iostream>
+#include <cmath>
+#include <algorithm>
+#include <thread>
+#include "utils/exception.h"
+
+#include "CL/OpenCLUtils.h"
+
+#include "CL/OpenCLParams.h"
+#include "CL/OpenCL.h"
+
+#ifdef __APPLE__
+#include <Accelerate/Accelerate.h>
+#endif
+#ifdef USE_MKL
+#include <mkl.h>
+#endif
+#ifdef USE_OPENBLAS
+#include <cblas.h>
+#endif
+
+
+namespace lczero {
+  
+  namespace {
+    
+    
+    static constexpr int NUM_VALUE_INPUT_PLANES = 32;
+    static constexpr int NUM_POLICY_INPUT_PLANES = 32;
+    
+    static constexpr int NUM_OUTPUT_POLICY = 1858;
+    static constexpr int NUM_VALUE_CHANNELS = 128;
+    
+    
+    class OpenCLNetwork;
+    
+    // Copy the vectors we need after weights is deallocated
+    struct OpenCLWeights {
+      
+      const std::vector<float> ip2_val_w;
+      const std::vector<float> ip2_val_b;
+      
+      OpenCLWeights(const Weights& weights):
+      ip2_val_w(weights.ip2_val_w),
+      ip2_val_b(weights.ip2_val_b)
+      {
+        
+      }
+
+    };
+    
+    
+    
+    class OpenCLComputation : public NetworkComputation {
+      
+    public:
+      
+      OpenCLComputation(const OpenCL_Network& opencl_net, const OpenCLWeights& weights):
+      opencl_net_(opencl_net),
+      weights_(weights),
+      input_data_(kInputPlanes*64),
+      value_data_(NUM_VALUE_CHANNELS),
+      policy_data_(),
+      q_value_(0) {
+        
+      }
+      
+      virtual ~OpenCLComputation() {
+        
+      }
+      
+      // Adds a sample to the batch.
+      void AddInput(InputPlanes&& input) override {
+        planes_.emplace_back(input);
+      }
+
+      
+    public:
+      
+     // Do the computation.
+      void ComputeBlocking() override {
+        
+        for (auto& sample : planes_)
+          ComputeBlocking(sample);
+
+      }
+      
+      
+      void ComputeBlocking(const InputPlanes &sample) {
+        
+        
+        int index=0;
+        for (const InputPlane& plane : sample) {
+          float value=plane.value;
+          const uint64_t one=1;
+          for (int i=0; i<64; i++)
+            input_data_[index++]=((plane.mask&(one<<i))==0 ) ? 0 : value;
+        }
+        
+        std::vector<float> policy_data(NUM_OUTPUT_POLICY);
+        opencl_net_.forward(input_data_, policy_data, value_data_);
+        
+        // Get the moves
+        softmax(policy_data, policy_data);
+        policy_data_.emplace_back(move(policy_data));
+        
+        // Now get the score
+        const std::vector<float>& ip2_val_w=weights_.ip2_val_w;
+        const std::vector<float>& ip2_val_b=weights_.ip2_val_b;
+        
+        double winrate=innerproduct(ip2_val_w, value_data_)+ip2_val_b[0];
```

Comment:
Why is this double, not float?

Reply:
The result on thanh(winrate) is casted to float (because it's a vector<float>)

       double winrate=innerproduct(ip2_val_w, value_data_)+ip2_val_b[0];
        q_value_.emplace_back(std::tanh(winrate));